### PR TITLE
updated term 'relapse' in disease_status_at_followup to 'relapse or r…

### PR DESCRIPTION
**Pull request by hknahal**
…ecurrence' after discussions with WG where it was agreed the two terms are synonymous. Also updated validation script and tests.

Related to: https://github.com/icgc-argo/argo-dictionary/issues/111

### CODE REVIEW

File: schemas/follow_up.json [hunk 1]
```diff
@@ -131,12 +131,12 @@
       "meta": {
         "core": true,
         "dependsOn": "follow_up.disease_status_at_followup",
-        "notes": "This field is required to be submitted if disease_status_at_followup indicates progression or relapse.",
+        "notes": "This field is required to be submitted if disease_status_at_followup indicates progression or relapse or recurrence.",
```
Comment (rosibaj) [inline schemas/follow_up.json:134]: would it be more grammatically correct to say `
-  indicates a state of progression, relapse, or recurrence`

Comment (rosibaj) [reply schemas/follow_up.json:134]: same comments for fields below

**Comment by rosibaj** [comment]
@hknahal i imported this dictionary version into the dev platform (45.0).

One question about the script: 

image

It appears that only the value `progression` requires values:
image

Shouldnt this also be the case for:
-  "Loco-regional progression",	         
-  "Relapse or recurrence",

File: schemas/follow_up.json (hunk 1 above)
Comment (hknahal) [reply schemas/follow_up.json:134]: Hmm, I thought I had fixed the script validation but maybe I missed this one. You're right, it should be required if it's "relapse or recurrence". I'll correct this.

**Comment by hknahal** [comment]
> @hknahal i imported this dictionary version into the dev platform (45.0).
> 
> One question about the script:
> 
> image
> 
> It appears that only the value `progression` requires values:
> image
> 
> Shouldnt this also be the case for:
> 
>     * "Loco-regional progression",
> 
>     * "Relapse or recurrence",

**Comment by hknahal** [comment]
> @hknahal i imported this dictionary version into the dev platform (45.0).
> 
> One question about the script:
> 
> image
> 
> It appears that only the value `progression` requires values:
> image
> 
> Shouldnt this also be the case for:
> 
>     * "Loco-regional progression",
> 
>     * "Relapse or recurrence",

It should allow `relapse or recurrence` as well but I'm not sure why. The script validation references two scripts (`requiredWhenProgression` and `requiredWhenRelapse`) - see [here](https://github.com/icgc-argo/argo-dictionary/blob/master/references/validationFunctions/follow_up/method_of_progression_status.js) but it looks like `requiredWhenProgression` takes precedence and `requiredWhenRelapse` is ignored even though it should check this as well (see [here](https://github.com/icgc-argo/argo-dictionary/pull/131/files#diff-84b516ff0dc0a83bec395ec05fdd044a)). I'll look into this a bit more.

```

**Comment by wajiha-oicr** [comment]
@rosibaj @hknahal  The "required if relapse" script was applied to the `relapse_interval` field, not the `relapse_type` field. ref (https://github.com/icgc-argo/argo-dictionary/issues/31). 

If `relapse_type` also is a required field if method of progression status is set to 'relapse or... recurrence', then the script needs to applied to that field as well. There is currently no valdiation for `relapse_type`

**Comment by wajiha-oicr** [comment]
Additionally, the way the error message currently works is that fields with multiple scripts, return the error message of the first script they fail on. That's why you only see one error message.

**Comment by hknahal** [comment]
I added the script validation for `relapse_type` but how do I make the script fail if the following happens?

**Comment by hknahal** [comment]
I added the script validation for relapse_type but my unit test is failing (see [here](https://github.com/icgc-argo/argo-dictionary/pull/131/files#diff-5e705d7c28171dd1bf18887e39805ff4R93)) in the following case:
`disease_status_at_followup` : `stable`
`relapse_type` : `Distant recurrence/metastasis`

**Comment by hknahal** [comment]
Thanks @wajiha-oicr . I added the script validation for relapse_type but my unit test is failing (see [here](https://github.com/icgc-argo/argo-dictionary/pull/131/files#diff-5e705d7c28171dd1bf18887e39805ff4R93)) in the following case:

`disease_status_at_followup` : `stable`
`relapse_type` : `Distant recurrence/metastasis`

How do I make the script fail if disease_status_at_followup is another value besides `relapse or recurrence` and `relapse_type` is provided?

**Comment by wajiha-oicr** [comment]
There's a few other scripts that show how something like that might be done. For example the '[ensured_deceased](https://github.com/icgc-argo/argo-dictionary/blob/master/references/validationFunctions/donor/ensureDeceased.js#L13)' script that performs a "value must be provided, **if and only if** xyz..." sort of validation. 

Is it just **_this_** particular field (`relapse_type`) that you want the " provide value if and only if the value is xyz..." condition on? Because if so I would recommend writing that code within its own script just for the `relapse_type.js` file. Otherwise, if this is something you want implemented for every field that relies on the `requiredWhenRelapse` validation, then you can modify that existing script.

**Comment by hknahal** [comment]
@wajiha-oicr I modified the `requiredWhenRelapse` and `requiredWhenProgression` scripts but the unit tests are failing on these two unit tests:

``` FAIL  tests/follow_up/requiredWhenRelapse.test.js
  ● Unit Tests for Disease Status Follow-Up › Tests for the relapse_interval field. › 
 Test 4 : Disease status is not relapse or recurrence, and relapse interval is not provided 
Expecting result.valid to be: true

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: undefined

      147 |             test.each(unitTests)('\n Test %# : %s \nExpecting result.valid to be: %s',(description,target,inputs) =>{
      148 |                 const scriptOutput = validation(inputs, inputs[name], name);
    > 149 |                 expect(scriptOutput.valid).toBe(target);
          |                                            ^
      150 |             })
      151 |         })
      152 |         

      at tests/follow_up/requiredWhenRelapse.test.js:149:44

 FAIL  tests/follow_up/requiredWhenProgression.test.js
  ● Unit Tests for Disease Status Follow-Up › Tests for the method_of_progression_status field. › 
 Test 4 : Disease status is partial remission, without method of progression 
Expecting result.valid to be: true

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: undefined

      136 |             test.each(unitTests)('\n Test %# : %s \nExpecting result.valid to be: %s',(description,target,inputs) =>{
      137 |                 const scriptOutput = validation(inputs, inputs[name], name);
    > 138 |                 expect(scriptOutput.valid).toBe(target);
          |                                            ^
      139 |             })
      140 |         })
      141 |         

      at tests/follow_up/requiredWhenProgression.test.js:138:44
```
I thought the way I had written the code would return a `True` result, so I'm not sure why it's failing. Any ideas?

**Comment by wajiha-oicr** [comment]
Ok, I'll take a look !

File: references/validationFunctions/follow_up/baseScripts/requiredWhenProgression.js [hunk 2]
```diff
@@ -7,20 +7,25 @@
...
 
         const isRequired = diseaseStatus.match(/(progression)$/);
+        const otherStatus = ["stable", "partial remission", "no evidence of disease", "complete remission"];
 
         // checks for a string just consisting of whitespace
         const checkforEmpty = (entry) => {return /^\s+$/g.test(decodeURI(entry).replace(/^"(.*)"$/, '$1'))};
 
         if (isRequired && (!$field || checkforEmpty($field))){
             result = {valid: false, message: `'${$name}' is a required field if 'disease_status_at_followup' is of type 'progression'.` }
         }
```
Comment (wajiha-oicr) [inline references/validationFunctions/follow_up/baseScripts/requiredWhenProgression.js:21]:
```
  else if (!isRequired && ! (!$field || checkforEmpty($field))){
            result = {valid: false, message: `'${$name}' cannot be provided ..etc etc` }
        }
```

Comment (wajiha-oicr) [reply references/validationFunctions/follow_up/baseScripts/requiredWhenProgression.js:21]: this should suffice, I don't think the other conditions are needed

Comment (wajiha-oicr) [reply references/validationFunctions/follow_up/baseScripts/requiredWhenProgression.js:21]: adding this to the requiredwhenrelapse script, removing the other else-ifs, seems to be making the tests pass

File: schemas/follow_up.json (hunk 1 above)
Comment (hknahal) [reply schemas/follow_up.json]: @rosibaj I fixed this.

**Pull request merged**